system/jlib/jqueue.hpp

```diff
@@ -0,0 +1,387 @@
+                }
+                else
+                {
+                    if ((curState & readerMask) != 0)
+                        nextState -= (1 << readerShift);
+                }
+                if (state.compare_exchange_weak(curState, nextState, std::memory_order_relaxed))
+                {
+                    unsigned slotMask = (fixedSlotBits ? fixedSlotMask : dynamicSlotMask);
+                    unsigned curDequeueSeq = (curState >> dequeueShift); // No need to mask since the top field
+                    unsigned curEnqueueSeq = (curDequeueSeq + curCount) & sequenceMask;
+                    unsigned filledSeq = (curEnqueueSeq + 1) & sequenceMask;
+                    unsigned curEnqueueSlot = curEnqueueSeq & slotMask;
+                    BufferElement & cur = values[curEnqueueSlot];
+
+                    //MORE: Another producer has been interrupted while writing to the same slot
+                    //or the consumer has not yet read from the slot.
+                    //spin until that has been consumed.
+                    unsigned spins = 0;
+                    while (cur.sequence.load(std::memory_order_acquire) != curEnqueueSeq)
+                    {
+                        if (slotUnavailableSpins != 0 && ++spins == slotUnavailableSpins)
+                        {
+                            ThreadYield();
+                            spins = 0;
+                        }
+                        else
+                            spinPause();
+                    }
+
+                    //enqueue takes ownership of the object -> use std::move
+                    cur.value = std::move(value);
+                    cur.sequence.store(filledSeq, std::memory_order_release);
+                    if ((curState & readerMask) != 0)
+                        readers.signal();
+                    return true;
+                }
+            }
+        }
+    }
+
+    bool dequeue(ELEMENT & result, bool returnIfEmpty)
+    {
+        if (aborted.load(std::memory_order_relaxed))
+            return false;
+
+        unsigned numSpins = initialSpinsBeforeWait;
+        //Note, compare_exchange_weak updates curState when it fails, so don't read inside the main loop
+        state_t curState = state.load(std::memory_order_acquire);
+        loop
+        {
+            unsigned curCount = (curState & countMask);
+            //Check if the queue is empty
+            if (curCount == 0)
+            {
+                if (returnIfEmpty)
+                    return false;
+
+                //If all writers have finished then no more items will be enqueued
+                if (allWritersStopped())
+                {
+                    curState = state.load(std::memory_order_acquire);
+
+                    //Check that nothing has been added since the previous load. (Very small window.)
+                    if ((curState & countMask) == 0)
+                        return false;
+                    continue;
```

Comment:
The aim is to consume any added items, or return.  If nothing has been added the return above will have been executed.  If something has it will take a different path in the main loop.
It is illegal for any writers to add items once allWritersStopped() is true - so it will only hit this continue once.